File: backend/app/schemas/salary.py
"""工资相关 Pydantic 模型：工资补贴/扣款、工资条、工资台账"""

from datetime import datetime

from pydantic import BaseModel, Field


# ── 工资补贴/扣款 ──

class SalaryAllowanceCreateIn(BaseModel):
    """工资补贴/扣款 — 创建"""
    user_id: int = Field(ge=1, description="员工 ID")
    allowance_type: str = Field(min_length=1, max_length=16, description="类型: bonus/deduction")
    amount: float = Field(description="金额")
    month: str = Field(min_length=7, max_length=7, description="月份 YYYY-MM")
    reason: str | None = Field(default=None, max_length=255, description="原因")


class SalaryAllowanceUpdateIn(BaseModel):
    """工资补贴/扣款 — 更新"""
    user_id: int | None = Field(default=None, ge=1, description="员工 ID")
    allowance_type: str | None = Field(default=None, max_length=16, description="类型: bonus/deduction")
    amount: float | None = Field(default=None, description="金额")
    month: str | None = Field(default=None, min_length=7, max_length=7, description="月份 YYYY-MM")
    reason: str | None = Field(default=None, max_length=255, description="原因")


class SalaryAllowanceOut(BaseModel):
    """工资补贴/扣款 — 输出"""
    id: int
    user_id: int
    allowance_type: str
    amount: float
    month: str
    reason: str | None
    created_at: datetime


# ── 工资条 ──

class SalarySlipOut(BaseModel):
    """工资条 — 输出（含计算字段）"""
    id: int
    user_id: int
    user_name: str | None = None
    month: str
    total_qty: int
    item_amount: float
    bonus_amount: float
    deduction_amount: float
    net_amount: float
    signature_attachment_id: int | None
    signed_at: datetime | None
    is_signed: bool
    confirm_status: str
    reject_reason: str | None
    rejected_at: datetime | None


class SalarySlipSignOut(BaseModel):
    """工资条签名后输出"""
    id: int
    month: str
    signature_attachment_id: int | None
    signed_at: datetime | None
    confirm_status: str


class SalarySlipRejectOut(BaseModel):
    """工资条拒签后输出"""
    id: int
    month: str
    confirm_status: str
    reject_reason: str | None
    rejected_at: datetime | None


class SalarySlipResetOut(BaseModel):
    """工资条重置确认后输出"""
    id: int
    confirm_status: str


# ── 工资台账 ──

class SalaryLedgerItemOut(BaseModel):
    """工资台账明细 — 输出"""
    id: int
    source: str
    salary_id: int | None
    report_unit_id: int | None
    report_id: int | None
    user_id: int
    username: str | None
    user_full_name: str | None
    order_id: int | None
    order_code: str | None
    product_id: int | None
    product_name: str | None
    sku_id: int | None
    sku_code: str | None
    sku_name: str | None
    process_id: int | None
    process_code: str | None
    process_name: str | None
    unit_seq: int | None
    reported_qty: int
    unit_price: float
    amount: float
    status: str
    status_label: str
    reported_at: datetime | None
    month: str
    task_code: str | None
    result_type: str | None

File: frontend-admin-pro/src/api/salary.ts
import { http } from '@/utils/http'
import type { ListResp } from '@/types/api'

export type SalaryItemOut = {
  id: number
  report_id: number | null
  user_id: number
  sku_id: number
  process_id: number
  unit_price: number
  good_qty: number
  amount: number
  month: string
  created_at: string
}

/** 工资明细台账（对标 thinkmes 工资列表） */
export type SalaryLedgerOut = {
  id: number
  source: 'unit' | 'report'
  salary_id: number | null
  report_unit_id: number | null
  report_id: number | null
  user_id: number
  username: string | null
  user_full_name: string | null
  order_id: number | null
  order_code: string | null
  product_id: number | null
  product_name: string | null
  sku_id: number | null
  sku_code: string | null
  sku_name: string | null
  process_id: number | null
  process_code: string | null
  process_name: string | null
  unit_seq: number | null
  reported_qty: number
  unit_price: number
  amount: number
  status: string
  status_label: string
  reported_at: string | null
  month: string
  task_code: string | null
  result_type: string | null
}

export type SalarySlipOut = {
  id: number
  user_id: number
  user_name: string | null
  month: string
  total_qty: number
  item_amount: number
  hourly_amount: number
  hourly_hours: number
  bonus_amount: number
  deduction_amount: number
  net_amount: number
  signature_attachment_id: number | null
  signed_at: string | null
  is_signed: boolean
  confirm_status?: string
  reject_reason?: string | null
  rejected_at?: string | null
}

export type HourlyItemOut = {
  id: number
  user_id: number
  user_name: string | null
  item_type: string
  work_date: string | null
  work_hours: number
  hourly_rate: number
  amount: number
  month: string
  is_absent: boolean
}

export type ExportJobOut = {
  id: number
  job_type: string
  status: string
  params: any
  result_attachment_id: number | null
  error_msg: string | null
  created_by: number | null
  created_at: string
  started_at: string | null
  finished_at: string | null
}

export const salaryApi = {
  // 工资
  listSalaryLedger(params?: {
    month?: string
    user_id?: number
    status?: string
    keyword?: string
    offset?: number
    limit?: number
  }) {
    return http.request<{ items: SalaryLedgerOut[]; total: number }>({
      url: '/admin/production/reports/salary/ledger',
      method: 'GET',
      params,
    })
  },
  listSalaryItems(params: any) {
    return http.request<ListResp<SalaryItemOut>>({ url: '/admin/production/reports/salary/items', method: 'GET', params })
  },
  getSalarySummary(params: any) {
    return http.request<{ items: any[] }>({ url: '/admin/production/reports/salary/summary', method: 'GET', params })
  },
  listSalaryAllowances(params?: { user_id?: number; month?: string }) {
    return http.request<{ items: { id: number; user_id: number; allowance_type: string; amount: number; month: string; reason: string | null }[] }>({
      url: '/admin/production/reports/salary/allowances',
      method: 'GET',
      params,
    })
  },
  createSalaryAllowance(data: {
    user_id: number
    allowance_type: string
    amount: number
    month: string
    reason?: string
  }) {
    return http.request<{ id: number; allowance_type: string; amount: number }>({
      url: '/admin/production/reports/salary/allowances',
      method: 'POST',
      data,
    })
  },
  exportSalaryExcel(params: any) {
    return http.request<Blob>({ url: '/admin/production/reports/salary/export', method: 'GET', params, responseType: 'blob' })
  },
  createSalaryExportJob(params: any) {
    return http.request<ExportJobOut>({ url: '/admin/production/reports/salary/export-jobs', method: 'POST', params })
  },
  getSalaryExportJob(id: number) {
    return http.request<ExportJobOut>({ url: `/admin/production/reports/salary/export-jobs/${id}`, method: 'GET' })
  },
  listSalarySlips(params: any) {
    return http.request<ListResp<SalarySlipOut>>({ url: '/admin/production/reports/salary/slips', method: 'GET', params })
  },
  resetSalarySlipConfirm(id: number) {
    return http.request<any>({ url: `/admin/production/reports/salary/slips/${id}/reset-confirm`, method: 'POST' })
  },
  remindSalarySlips(month: string) {
    return http.request<{ month: string; sent: number; skipped: number }>({
      url: '/admin/production/reports/salary/slips/remind',
      method: 'POST',
      params: { month },
    })
  },

  // 计时工资
  listHourlyItems(params?: { month?: string; user_id?: number; offset?: number; limit?: number }) {
    return http.request<{ items: HourlyItemOut[]; total: number }>({
      url: '/admin/production/reports/salary/hourly-items',
      method: 'GET',
      params,
    })
  },
  generateTimeItems(params: { date_from: string; date_to?: string; user_id?: number }) {
    return http.request<{ generated: number; date_from: string; date_to: string }>({
      url: '/admin/production/reports/salary/generate-time-items',
      method: 'POST',
      params,
    })
  },
  getHourlySummary(params?: { month?: string; user_id?: number }) {
    return http.request<{ total_hours: number; total_amount: number }>({
      url: '/admin/production/reports/salary/hourly-summary',
      method: 'GET',
      params,
    })
  },
  listHourlyLedger(params?: { month?: string; user_id?: number; offset?: number; limit?: number }) {
    return http.request<{ items: any[]; total: number }>({
      url: '/admin/production/reports/salary/hourly-ledger',
      method: 'GET',
      params,
    })
  },
}
